services/py-api/src/service/mail_service/resend_service.py

```diff
@@ -0,0 +1,95 @@
+import os
+from typing import Optional
+from dotenv import load_dotenv
+
+import resend
+from mail_service import MailService
+from service.mail_service.utils import load_email_participant_html_template, load_email_verify_participant_html_template
+
+SENDER_EMAIL = "tonymontana@thehub.com"  # We obviously need to supply our own email here.
+load_dotenv()
+
+
+class ResendMailService(MailService):
+    def __init__(self) -> None:
+        api_key = os.getenv("RESEND_API_KEY")
+        if not api_key:
+            raise ValueError("RESEND_API_KEY environment variable is not set")
+        resend.api_key = api_key
+        self.client = resend
+
+    async def send_email(
+        self, receiver: str, subject: str, body_text: Optional[str] = None, body_html: Optional[str] = None
+    ) -> None:
+        if not receiver:
+            raise ValueError("Receiver email must be provided.")
+        if not subject:
+            raise ValueError("Subject must be provided.")
+
+        if not body_text and not body_html:
+            raise ValueError("At least one of body_text or body_html must be provided")
+
+        params: resend.Emails.SendParams = {
+            "from": f"THE HUB <{SENDER_EMAIL}>",
+            "to": [receiver],
+            "subject": subject,
+        }
+
+        if body_html:
+            params["html"] = body_html
+        elif body_text:
+            params["text"] = body_text
+
+        try:
+            email = self.client.Emails.send(params)
```

Comment:
Running a sync method inside an async ctx will block the event loop, and we will not be able to process any requests until the task is completed. The tasks that run in an async ctx should be able to yield control back to the event loop. How to avoid this situation:

1. Preferred approach: remove the async and write in the docs that this method is blocking and should be executed via [BackgroundTasks.add_task()](https://fastapi.tiangolo.com/tutorial/background-tasks/)

2. BackgroundTasks uses [to_thread](https://docs.python.org/3/library/asyncio-task.html#asyncio.to_thread) under the hood so you can write the wrapper yourself to send the email asynchronously. 

3. If, for some reason, we need to have a cancelation mechanism due to some deadline being exceeded (timeout) via [wait_for](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for), the first and second approaches will not be useful. The wait_for will cancel the future returned (not the thread) but unfortunately the thread will continue running indefinitely, which could create strange behavior. You can check more [here](https://github.com/python/cpython/issues/87185). Instead we should be using [asyncio.loop.run_in_executor](https://docs.python.org/3.10/library/asyncio-eventloop.html#asyncio.loop.run_in_executor) to schedule the background task and get a future we can use `wait_for` to wait on the event-loop.

Something else, I am not sure if `send` has a retry logic underneath, but as we are working with remote systems is important to have a retry logic on TransientErrors or 500 response codes. You can grab some inspiration from here: https://github.com/AUBGTheHUB/monolith/blob/30fe96921c1b8f212274389057d5bc39b5aed654/services/py-api/src/database/transaction_manager.py#L43

Reply:
Why would it block the event loop. I think he will only be left with email having an undefined value since it did not wait for the request to finish. Then if we are calling this async 'send_mail' method from a syncronous context we need to use the BackgroundTasks, otherwise we are blocking the flow of the application @IvanObreshkov

Reply:
Yes @Isak-Bego, @cl3vy's version of the code right now does NOT use `BackgroundTasks` and is calling the sync method `resend.Emails.send` inside a function that is declared to be async. If the caller calls this function  as it is right now and awaits it, the whole event loop will be blocked.

Reply:
Answering this:

<img width="615" alt="image" src="https://github.com/user-attachments/assets/a7366ac0-0410-4560-8a70-4ddaaf4cae08">


WELL `resend.Emails.send(params)` IS NOT ASYNC, this why it is not awaited :D
<img width="512" alt="image" src="https://github.com/user-attachments/assets/24db63bf-ed87-4cbf-80ac-fd686b2d2f04">

Reply:
> Also, I guess you have checked the API response codes. Currently your code does not handle this case: https://resend.com/docs/api-reference/introduction#rate-limit

One good thing the resend SDK does is that it takes care of this.